File: NextcloudTalk/Calls/NCAudioController.m
/**
 * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import "NCAudioController.h"

#import "CallKitManager.h"

#import "NextcloudTalk-Swift.h"

NSString * const AudioSessionDidChangeRouteNotification             = @"AudioSessionDidChangeRouteNotification";
NSString * const AudioSessionWasActivatedByProviderNotification     = @"AudioSessionWasActivatedByProviderNotification";
NSString * const AudioSessionDidChangeRoutingInformationNotification   = @"AudioSessionDidChangeRoutingInformationNotification";

@implementation NCAudioController

+ (NCAudioController *)sharedInstance
{
    static NCAudioController *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[NCAudioController alloc] init];
    });
    return sharedInstance;
}

- (id)init
{
    self = [super init];
    if (self) {
        RTCAudioSessionConfiguration *configuration = [RTCAudioSessionConfiguration webRTCConfiguration];
        configuration.category = AVAudioSessionCategoryPlayAndRecord;
        configuration.mode = AVAudioSessionModeVoiceChat;
        [RTCAudioSessionConfiguration setWebRTCConfiguration:configuration];
        
        _rtcAudioSession = [RTCAudioSession sharedInstance];
        [_rtcAudioSession lockForConfiguration];
        NSError *error = nil;
        [_rtcAudioSession setConfiguration:configuration error:&error];
        if (error) {
            NSLog(@"Error setting configuration: %@", error.localizedDescription);
        }
        [_rtcAudioSession unlockForConfiguration];
        
        if ([CallKitManager isCallKitAvailable]) {
            _rtcAudioSession.useManualAudio = YES;
        }
        
        [_rtcAudioSession addDelegate:self];

        [self updateRouteInformation];
    }
    return self;
}

#pragma mark - Audio session configuration

- (void)setAudioSessionToVoiceChatMode
{
    [self changeAudioSessionConfigurationModeTo:AVAudioSessionModeVoiceChat];
}

- (void)setAudioSessionToVideoChatMode
{
    [self changeAudioSessionConfigurationModeTo:AVAudioSessionModeVideoChat];
}

- (void)changeAudioSessionConfigurationModeTo:(NSString *)mode
{
    [[WebRTCCommon shared] assertQueue];

    RTCAudioSessionConfiguration *configuration = [RTCAudioSessionConfiguration webRTCConfiguration];
    configuration.category = AVAudioSessionCategoryPlayAndRecord;
    configuration.mode = mode;

    [_rtcAudioSession lockForConfiguration];
    BOOL hasSucceeded = NO;
    NSError *error = nil;

    if (_rtcAudioSession.isActive) {
        hasSucceeded = [_rtcAudioSession setConfiguration:configuration error:&error];
    } else {
        hasSucceeded = [_rtcAudioSession setConfiguration:configuration
                                          active:YES
                                           error:&error];
    }

    if (!hasSucceeded) {
        NSLog(@"Error setting configuration: %@", error.localizedDescription);
    }

    [_rtcAudioSession unlockForConfiguration];

    [self updateRouteInformation];
}

- (void)disableAudioSession
{
    [[WebRTCCommon shared] assertQueue];

    [_rtcAudioSession lockForConfiguration];

    NSError *error = nil;
    BOOL hasSucceeded = [_rtcAudioSession setActive:NO error:&error];

    if (!hasSucceeded) {
        NSLog(@"Error setting configuration: %@", error.localizedDescription);
    }

    [_rtcAudioSession unlockForConfiguration];
}

- (void)updateRouteInformation
{
    AVAudioSession *audioSession = self.rtcAudioSession.session;
    AVAudioSessionPortDescription *currentOutput = [audioSession.currentRoute.outputs firstObject];

    self.numberOfAvailableInputs = audioSession.availableInputs.count;

    if ([_rtcAudioSession mode] == AVAudioSessionModeVideoChat || [currentOutput.portType isEqualToString:AVAudioSessionPortBuiltInSpeaker]) {
        self.isSpeakerActive = YES;
    } else {
        self.isSpeakerActive = NO;
    }

    [[NSNotificationCenter defaultCenter] postNotificationName:AudioSessionDidChangeRoutingInformationNotification
                                                        object:self
                                                      userInfo:nil];
}

- (BOOL)isAudioRouteChangeable
{
    if (self.numberOfAvailableInputs > 1) {
        return YES;
    }

    if ([UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPhone) {
        // A phone always supports a speaker and earpiece output
        return YES;
    }

    return NO;
}

- (void)providerDidActivateAudioSession:(AVAudioSession *)audioSession
{
    [[WebRTCCommon shared] assertQueue];

    [_rtcAudioSession audioSessionDidActivate:audioSession];
    _rtcAudioSession.isAudioEnabled = YES;

    [[NSNotificationCenter defaultCenter] postNotificationName:AudioSessionWasActivatedByProviderNotification
                                                        object:self
                                                      userInfo:nil];
}

- (void)providerDidDeactivateAudioSession:(AVAudioSession *)audioSession
{
    [[WebRTCCommon shared] assertQueue];

    [_rtcAudioSession audioSessionDidDeactivate:audioSession];
    _rtcAudioSession.isAudioEnabled = NO;
}

#pragma mark - RTCAudioSessionDelegate

- (void)audioSessionDidChangeRoute:(RTCAudioSession *)session reason:(AVAudioSessionRouteChangeReason)reason previousRoute:(AVAudioSessionRouteDescription *)previousRoute
{
    [self updateRouteInformation];
    
    [[NSNotificationCenter defaultCenter] postNotificationName:AudioSessionDidChangeRouteNotification
                                                        object:self
                                                      userInfo:nil];
}

@end

File: NextcloudTalk/Calls/NCAudioController.swift
//
// SPDX-FileCopyrightText: 2022 Nextcloud GmbH and Nextcloud contributors
// SPDX-License-Identifier: GPL-3.0-or-later
//

import WebRTC

@objcMembers
public class NCAudioController: NSObject, RTCAudioSessionDelegate {

    public static let shared = NCAudioController()

    public var rtcAudioSession: RTCAudioSession
    public var isSpeakerActive: Bool = false
    public var numberOfAvailableInputs: Int = 0

    public var isAudioRouteChangeable: Bool {
        if self.numberOfAvailableInputs > 1 {
            return true
        }

        if UIDevice.current.userInterfaceIdiom == .phone {
            // A phone always supports a speaker and earpiece output
            return true
        }

        return false
    }

    public override init() {
        let configuration = RTCAudioSessionConfiguration.webRTC()
        configuration.category = AVAudioSession.Category.playAndRecord.rawValue
        configuration.mode = AVAudioSession.Mode.voiceChat.rawValue
        RTCAudioSessionConfiguration.setWebRTC(configuration)

        self.rtcAudioSession = RTCAudioSession.sharedInstance()
        self.rtcAudioSession.lockForConfiguration()

        do {
            try self.rtcAudioSession.setConfiguration(configuration)
        } catch {
            NCUtils.log("Error setting audio configuration: \(error.localizedDescription)")
        }

        self.rtcAudioSession.unlockForConfiguration()

        if CallKitManager.isCallKitAvailable() {
            self.rtcAudioSession.useManualAudio = true
        }

        super.init()

        self.rtcAudioSession.add(self)

        self.updateRouteInformation()
    }

    // MARK: - Audio session configuration

    public func setAudioSessionToVoiceChatMode() {
        self.changeAudioSessionConfiguration(toMode: .voiceChat)
    }

    public func setAudioSessionToVideoChatMode() {
        self.changeAudioSessionConfiguration(toMode: .videoChat)
    }

    private func changeAudioSessionConfiguration(toMode newMode: AVAudioSession.Mode) {
        WebRTCCommon.shared.assertQueue()

        let configuration = RTCAudioSessionConfiguration.webRTC()
        configuration.category = AVAudioSession.Category.playAndRecord.rawValue
        configuration.mode = newMode.rawValue

        self.rtcAudioSession.lockForConfiguration()

        do {
            if self.rtcAudioSession.isActive {
                try self.rtcAudioSession.setConfiguration(configuration)
            } else {
                try self.rtcAudioSession.setConfiguration(configuration, active: true)
            }
        } catch {
            NCUtils.log("Error setting audio configuration: \(error.localizedDescription)")
        }

        self.rtcAudioSession.unlockForConfiguration()

        self.updateRouteInformation()
    }

    public func disableAudioSession() {
        WebRTCCommon.shared.assertQueue()

        self.rtcAudioSession.lockForConfiguration()

        do {
            try self.rtcAudioSession.setActive(false)
        } catch {
            NCUtils.log("Error setting audio configuration: \(error.localizedDescription)")
        }

        self.rtcAudioSession.unlockForConfiguration()
    }

    private func updateRouteInformation() {
        let audioSession = self.rtcAudioSession.session
        let currentOutput = audioSession.currentRoute.outputs.first

        self.numberOfAvailableInputs = audioSession.availableInputs?.count ?? 0

        if self.rtcAudioSession.mode == AVAudioSession.Mode.videoChat.rawValue || currentOutput?.portType == .builtInSpeaker {
            self.isSpeakerActive = true
        } else {
            self.isSpeakerActive = false
        }

        NotificationCenter.default.post(name: .AudioSessionDidChangeRoutingInformation, object: self)
    }

    public func providerDidActivate(audioSession: AVAudioSession) {
        WebRTCCommon.shared.assertQueue()

        self.rtcAudioSession.audioSessionDidActivate(audioSession)
        self.rtcAudioSession.isAudioEnabled = true

        NotificationCenter.default.post(name: .AudioSessionWasActivatedByProvider, object: self)
    }

    public func providerDidDeactivate(audioSession: AVAudioSession) {
        WebRTCCommon.shared.assertQueue()

        self.rtcAudioSession.audioSessionDidDeactivate(audioSession)
        self.rtcAudioSession.isAudioEnabled = false
    }

    // MARK: RTCAudioSessionDelegate

    public func audioSessionDidChangeRoute(_ session: RTCAudioSession, reason: AVAudioSession.RouteChangeReason, previousRoute: AVAudioSessionRouteDescription) {
        self.updateRouteInformation()

        NotificationCenter.default.post(name: .AudioSessionDidChangeRoute, object: self)
    }

}
